### PR TITLE
fix(server): Set content-length and use faster fetch API for file transfers

### LINKED ISSUE #2798: Web downloads capped at 7.5GB
### What went wrong?

Downloading a 13GB file https://openneuro.org/crn/datasets/ds002179/snapshots/1.1.0/files/sub-EXC004:anat:sub-EXC004_acq-FA15_FLASH.nii.gz "completes" at 7.5GB:

```console
$ wget https://openneuro.org/crn/datasets/ds002179/snapshots/1.1.0/files/sub-EXC004:anat:sub-EXC004_acq-FA15_FLASH.nii.gz
--2023-04-12 07:29:28--  https://openneuro.org/crn/datasets/ds002179/snapshots/1.1.0/files/sub-EXC004:anat:sub-EXC004_acq-FA15_FLASH.nii.gz
Resolving openneuro.org (openneuro.org)... 34.107.158.10
Connecting to openneuro.org (openneuro.org)|34.107.158.10|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/gzip]
Saving to: ‘sub-EXC004:anat:sub-EXC004_acq-FA15_FLASH.nii.gz’

sub-EXC004:anat:sub-EXC0     [       <=>                         ]   7.43G  18.9MB/s    in 6m 12s  

2023-04-12 07:35:40 (20.4 MB/s) - ‘sub-EXC004:anat:sub-EXC004_acq-FA15_FLASH.nii.gz’ saved [7974987415]
```

Fetching the dataset through DataLad or S3 retrieves the whole file.

### Expected behavior

Either serve the whole file or refuse to serve.

### How to reproduce

```
$ wget https://openneuro.org/crn/datasets/ds002179/snapshots/1.1.0/files/sub-EXC004:anat:sub-EXC004_acq-FA15_FLASH.nii.gz
```

### Desktop

- OS: Ubuntu
- Browser: n/a
- Version:

### Phone

- Device:
- OS:
- Browser:
- Version:

### Additional information

xref freshdesk#1459

### PR DESCRIPTION
**Pull request by nellh**
This replaces the superagent variant of getFile with the fetch equivalent. Testing locally with a 10GB nifti this is a little faster than the superagent version, which had been used for performance reasons originally.

```shell
# Superagent
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  9.8G    0  9.8G    0     0   975M      0 --:--:--  0:00:10 --:--:--  959M
# Fetch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  9.8G    0  9.8G    0     0  1392M      0 --:--:--  0:00:07 --:--:-- 1412M
```

This change also adds a content-length header to the response, allowing an HTTP client to detect errors when the connection is closed early.

This may fix #2798, I think this is worth trying in staging (it isn't reproducible locally).

### CODE REVIEW

**Comment by effigies** [comment]
No idea if this is sensible or plausible, but would it work to redirect to the S3 URL for snapshots? (Will look at this PR tomorrow.)

**Comment by nellh** [comment]
> No idea if this is sensible or plausible, but would it work to redirect to the S3 URL for snapshots? (Will look at this PR tomorrow.)

I think I have a quick improvement for that in mind, I'll make that separately.

File: packages/openneuro-server/src/handlers/datalad.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import request from 'superagent'
```
Comment (effigies) [inline packages/openneuro-server/src/handlers/datalad.js:1]:
```suggestion
```

File: packages/openneuro-server/src/handlers/datalad.js [hunk 2]
```diff
@@ -19,9 +20,22 @@ export const getFile = (req, res) => {
   const worker = getDatasetWorker(datasetId)
   res.set('Content-Type', mime.lookup(filename) || 'application/octet-stream')
   const uri = snapshotId
-    ? `${worker}/datasets/${datasetId}/snapshots/${snapshotId}/files/${filename}`
-    : `${worker}/datasets/${datasetId}/files/${filename}`
-  return request.get(uri).pipe(res)
+    ? `http://${worker}/datasets/${datasetId}/snapshots/${snapshotId}/files/${filename}`
+    : `http://${worker}/datasets/${datasetId}/files/${filename}`
```
Comment (effigies) [inline packages/openneuro-server/src/handlers/datalad.js:24]: Is this a change demanded by `fetch()` vs `request()`, or just an "explicit is better than implicit" sort of thing?

**Review by effigies — APPROVED** [review]
This reads sensibly. If I understand it correctly, it takes an external URL, translates it to an internal URL and then proxies the internal stream back out to the external request.

Probably want to remove L11-L12 comment.

File: packages/openneuro-server/src/handlers/datalad.js (hunk 1 above)
Comment (nellh) [reply packages/openneuro-server/src/handlers/datalad.js:1]: It's still used later in the same module. Might be a good idea to drop it entirely, it's from an earlier era in JavaScript request libraries, but still needed here.

File: packages/openneuro-server/src/handlers/datalad.js (hunk 2 above)
Comment (nellh) [reply packages/openneuro-server/src/handlers/datalad.js:24]: Fetch requires the protocol, it's implicit with superagent.

File: packages/openneuro-server/src/handlers/datalad.js (hunk 1 above)
Comment (effigies) [reply packages/openneuro-server/src/handlers/datalad.js:1]: Ah, I missed that there was another function later in the file. GitHub clipped convincingly.

**Pull request merged**